### PR TITLE
Component library

### PR DESCRIPTION
**Pull request by joshdschneider**

### CODE REVIEW

File: src/components/Mfa.tsx [hunk 1]
```diff
@@ -0,0 +1,354 @@
...
+            const res = await mfaApi.mfaEnable({ code })
+            if (res.ok) {
+                setShowEnableModal(false)
+                setCode("")
+                setError(undefined)
+                setMfaStatus("Enabled")
+            } else {
+                res.error._visit({
+                    notFoundMfaEnable: () => setError(NOT_FOUND_MFA_ENABLE),
+                    badRequestMfaEnable: () => setError(BAD_REQUEST_MFA_ENABLE),
```
Comment (andrew-propelauth) [inline src/components/Mfa.tsx:126]: Same badRequest comment

File: src/components/Mfa.tsx [hunk 2]
```diff
@@ -0,0 +1,354 @@
...
+        try {
+            setLoading(true)
+            const res = await mfaApi.mfaDisable()
+            if (res.ok) {
+                setShowDisableModal(false)
+                setError(undefined)
+                setMfaStatus("Disabled")
+            } else {
+                res.error._visit({
+                    notFoundMfaDisable: () => setError(NOT_FOUND_MFA_DISABLE),
```
Comment (andrew-propelauth) [inline src/components/Mfa.tsx:150]: This error message doesn't reflect what's happening, I'll update the API

File: src/components/Mfa.tsx [hunk 3]
```diff
@@ -0,0 +1,354 @@
...
+            document.body.removeChild(link)
+            window.URL.revokeObjectURL(url)
+        } catch (e) {
+            setError("Download failed")
+            console.error(e)
+        }
+    }
+
+    if (mfaStatus === "Loading" || mfaStatus === "Error") {
+        return null // ?
```
Comment (andrew-propelauth) [inline src/components/Mfa.tsx:182]: Loading probably needs it's own state?

For error, we should have some generic "An unexpected error has occurred, please try again"

File: src/components/SignInOptions.tsx [hunk 4]
```diff
@@ -0,0 +1,176 @@
...
+export const SignInOptions = ({ config, buttonAppearance }: SignInOptionsProps) => {
+    const GOOGLE_LOGIN_PATH = "/google/login"
+    const GITHUB_LOGIN_PATH = "/github/login"
+    const SLACK_LOGIN_PATH = "/slack/login"
+    const MICROSOFT_LOGIN_PATH = "/microsoft/login"
+    const LINKEDIN_LOGIN_PATH = "/linkedin/login"
+    const PASSWORDLESS_LOGIN_PATH = "/login_passwordless"
+
+    const loginWithSocial = (path: string) => {
+        const url = path // use url from config?
```
Comment (andrew-propelauth) [inline src/components/SignInOptions.tsx:20]: Yeah, for loginWithSocial this should be the authUrl followed by the paths above (Google => LinkedIn). 

For passwordless, given that that's a component now, we'll need to load that from the config.

File: src/components/Signup.tsx [hunk 5]
```diff
@@ -0,0 +1,252 @@
...
+                {config && config.has_password_login && (
+                    <SignupForm config={config} onSuccess={onSuccess} presetEmail={presetEmail} />
+                )}
+                <BottomLinks onRedirectToLogin={onRedirectToLogin} appearance={appearance} />
+            </Container>
+        </div>
+    )
+}
+
+export type SignupOptions = {
```
Comment (andrew-propelauth) [inline src/components/Signup.tsx:92]: Doesn't this come from the API?

File: src/components/Signup.tsx [hunk 6]
```diff
@@ -0,0 +1,252 @@
...
+            const inviteToken = getTokenFromURL()
+            if (inviteToken) {
+                options.t = inviteToken
+            }
+            const response = await userApi.signup(options)
+            if (response.ok) {
+                onSuccess()
+            } else {
+                response.error._visit({
+                    badRequestSignup: () => setError(BAD_REQUEST_SIGNUP),
```
Comment (andrew-propelauth) [inline src/components/Signup.tsx:147]: Same badRequest* comment

File: src/components/UpdateProfile.tsx [hunk 7]
```diff
@@ -0,0 +1,378 @@
...
+        try {
+            event.preventDefault()
+            setLoading(true)
+            const res = await userApi.updateEmail({ newEmail: email })
+            if (res.ok) {
+                setShowConfirmationModal(true)
+            } else {
+                res.error._visit({
+                    notFoundUpdateEmail: () => setError(NOT_FOUND_UPDATE_EMAIL),
+                    badRequestUpdateEmail: () => setError(BAD_REQUEST_UPDATE_EMAIL),
```
Comment (andrew-propelauth) [inline src/components/UpdateProfile.tsx:104]: Same badRequest* comment

**Review by andrew-propelauth — CHANGES_REQUESTED** [review]
Great work!

File: src/components/Login.tsx [hunk 8]
```diff
@@ -0,0 +1,217 @@
...
+import { SignInOptions } from "./SignInOptions"
+import { UserMetadata, UserMetadataAppearance } from "./UserMetadata"
+import { Verify, VerifyAppearance } from "./Verify"
+
+export type LoginProps = {
+    onSuccess: VoidFunction
+    onRedirectToSignup?: VoidFunction
+    onRedirectToForgotPassword?: VoidFunction
+    presetEmail?: string
+    appearance?: LoginAppearance
```
Comment (joshdschneider) [inline src/components/Login.tsx:27]: Join types

File: src/AuthContext.tsx [hunk 9]
```diff
@@ -21,10 +24,14 @@ interface InternalAuthState {
     redirectToCreateOrgPage: () => void
 
     activeOrgFn: () => string | null
+
+    api: PropelAuthFeV2Client
 }
 
 export type AuthProviderProps = {
     authUrl: string
+    elements: Elements
```
[the remark this answers is not in the document]
Comment (joshdschneider) [reply src/AuthContext.tsx]: [fix](https://github.com/PropelAuth/react/commits/release/v2.1.0-beta.1#:~:text=optional%20elements%2C%20conditional%20wrappers)

File: src/AuthContext.tsx [hunk 10]
```diff
@@ -166,8 +175,15 @@ export const AuthProvider = (props: AuthProviderProps) => {
         redirectToOrgPage,
         redirectToCreateOrgPage,
         activeOrgFn,
+        api,
     }
-    return <AuthContext.Provider value={value}>{props.children}</AuthContext.Provider>
+    return (
+        <AuthContext.Provider value={value}>
+            <ElementsProvider elements={props.elements}>
```
[the remark this answers is not in the document]
Comment (joshdschneider) [reply src/AuthContext.tsx]: [fix](https://github.com/PropelAuth/react/commits/release/v2.1.0-beta.1#:~:text=appearance%20context%20error%20support)

File: src/components/ConfirmEmail.tsx [hunk 11]
```diff
@@ -0,0 +1,57 @@
...
+        Container?: ElementAppearance<ContainerProps>
+        Logo?: ElementAppearance<ImageProps>
+        Header?: ElementAppearance<H3Props>
+        Text?: ElementAppearance<ParagraphProps>
+    }
+}
+
+export const ConfirmEmail = ({ appearance }: ConfirmEmailProps) => {
+    const { config } = useConfig()
+    const message = `You should receive an email with a link to confirm your email address within the next few minutes. If you do not receive an email, make sure to check your spam.`
```
[the remark this answers is not in the document]
Comment (joshdschneider) [reply src/components/ConfirmEmail.tsx]: [fix](https://github.com/PropelAuth/react/pull/11/commits#:~:text=moved%20string%20to%20constants)

File: src/components/CreateOrg.tsx [hunk 12]
```diff
@@ -0,0 +1,147 @@
...
+import { H3, H3Props } from "../elements/H3"
+import { Image, ImageProps } from "../elements/Image"
+import { Input, InputProps } from "../elements/Input"
+import { Label } from "../elements/Label"
+import { useApi } from "../useApi"
+import { Config } from "../useConfig"
+import { BAD_REQUEST_CREATE_ORG, NOT_FOUND_CREATE_ORG, UNAUTHORIZED_ORG_USAGE, UNEXPECTED_ERROR } from "./constants"
+
+export type CreateOrgProps = {
+    config: Config | null
```
[the remark this answers is not in the document]
Comment (joshdschneider) [reply src/components/CreateOrg.tsx:17]: [fix](https://github.com/PropelAuth/react/pull/11/commits#:~:text=pull%20config%20from%20hook)

File: src/components/Login.tsx [hunk 13]
```diff
@@ -0,0 +1,217 @@
...
+            return <Verify setStep={setStep} appearance={verifyAppearance} />
+
+        case LoginStateEnum.UserMetadataRequired:
+            return <UserMetadata setStep={setStep} config={config} appearance={userMetadataAppearance} />
+
+        case LoginStateEnum.OrgCreationRequired:
+            return <CreateOrg setStep={setStep} config={config} appearance={createOrgAppearance} />
+
+        default:
+            return null // ?
```
[the remark this answers is not in the document]
Comment (joshdschneider) [reply src/components/Login.tsx]: [fix](https://github.com/PropelAuth/react/pull/11/commits#:~:text=org%20creation%20callback%2C%20login%20error%20fallback)

File: src/components/CreateOrg.tsx [hunk 14]
```diff
@@ -0,0 +1,147 @@
...
+    setStep: Dispatch<SetStateAction<LoginStateEnum>>
```
[the remark this answers is not in the document]
Comment (joshdschneider) [reply src/components/CreateOrg.tsx]: [fix](https://github.com/PropelAuth/react/pull/11/commits#:~:text=org%20creation%20callback%2C%20login%20error%20fallback)

File: src/components/CreateOrg.tsx [hunk 15]
```diff
@@ -0,0 +1,147 @@
...
+    appearance?: CreateOrgAppearance
+}
+
+export type CreateOrgAppearance = {
+    options?: {
+        headerContent?: ReactNode
+        displayLogo?: boolean
+        orgNameLabel?: ReactNode
+        autojoinByDomainLabel?: string
```
[the remark this answers is not in the document]
Comment (joshdschneider) [reply src/components/CreateOrg.tsx]: [fix](https://github.com/PropelAuth/react/pull/11/commits#:~:text=label%20is%20node%20not%20string)

File: src/AppearanceProvider.tsx [hunk 16]
```diff
@@ -0,0 +1,86 @@
...
+import { SelectProps } from "./elements/Select"
+import { TableProps } from "./elements/Table"
+
+export type Element<T> = (props: T) => JSX.Element
+
+export type ElementAppearance<T> = string | CSSProperties | Element<T>
+
+export type Appearance = {
+    options?: {
+        displayLogo?: boolean
```
[the remark this answers is not in the document]
Comment (joshdschneider) [reply src/AppearanceProvider.tsx]: [fix](https://github.com/PropelAuth/react/pull/11/commits#:~:text=removed%20global%20appearance%20options)

File: src/AppearanceProvider.tsx [hunk 17]
```diff
@@ -0,0 +1,86 @@
...
+                setAppearance: _setAppearance,
+            }}
+        >
+            {children}
+        </AppearanceContext.Provider>
+    )
+}
+
+export const useAppearance = () => {
+    const { appearance, setAppearance } = useContext(AppearanceContext)
```
[the remark this answers is not in the document]
Comment (joshdschneider) [reply src/AppearanceProvider.tsx]: [fix](https://github.com/PropelAuth/react/pull/11/commits#:~:text=appearance%20context%20error%20support)

File: src/components/CreateOrg.tsx [hunk 18]
```diff
@@ -0,0 +1,147 @@
...
+}
+
+export const CreateOrg = ({ config, setStep, appearance }: CreateOrgProps) => {
+    const { orgApi, loginApi } = useApi()
+    const [loading, setLoading] = useState(false)
+    const [name, setName] = useState("")
+    const [autojoinByDomain, setAutojoinByDomain] = useState(false)
+    const [restrictToDomain, setRestrictToDomain] = useState(false)
+    const [error, setError] = useState<string | undefined>(undefined)
+    const orgMetaname = (config && config.orgs_metaname) || "Organization"
```
[the remark this answers is not in the document]
Comment (joshdschneider) [reply src/components/CreateOrg.tsx]: [fix](https://github.com/PropelAuth/react/pull/11/commits/441f75146530ae01d13116328da98ec15a680e63)

File: src/components/ForgotPassword.tsx [hunk 19]
```diff
@@ -0,0 +1,222 @@
...
+        SubmitButton?: ElementAppearance<ButtonProps>
+        MagicLinkButton?: ElementAppearance<ButtonProps>
+        LoginButton?: ElementAppearance<ButtonProps>
+        ErrorMessage?: ElementAppearance<AlertProps>
+    }
+}
+
+export const ForgotPassword = ({ onRedirectToLogin, appearance }: ForgotPasswordProps) => {
+    const { loginApi } = useApi()
+    const { config } = useConfig()
```
[the remark this answers is not in the document]
Comment (joshdschneider) [reply src/components/ForgotPassword.tsx]: [fix](https://github.com/PropelAuth/react/pull/11/commits#:~:text=added%20config%20loading%20state)

File: src/components/ForgotPassword.tsx [hunk 20]
```diff
@@ -0,0 +1,222 @@
...
+}
+
+type ForgotPasswordDirectionsProps = {
+    appearance?: ForgotPasswordAppearance
+    hasPasswordlessLogin: boolean
+}
+
+const ForgotPasswordDirections = ({ appearance, hasPasswordlessLogin }: ForgotPasswordDirectionsProps) => {
+    const passwordMessage = `Enter your email address and we will send you an email with a link that will let you reset your password.`
+    const passwordlessMessage = `You can choose between receiving an email to reset your password or receiving an email with a magic link that will log you in.`
```
[the remark this answers is not in the document]
Comment (joshdschneider) [reply src/components/ForgotPassword.tsx]: [fix](https://github.com/PropelAuth/react/pull/11/commits#:~:text=moved%20password%20strings%20to%20constants)

File: src/components/InviteUser.tsx [hunk 21]
```diff
@@ -0,0 +1,128 @@
...
+        } catch (e) {
+            setError(UNEXPECTED_ERROR)
+            console.error(e)
+        } finally {
+            setLoading(false)
+        }
+    }
+
+    if (!inviteePossibleRoles || inviteePossibleRoles.length === 0) {
+        return null // ?
```
[the remark this answers is not in the document]
Comment (joshdschneider) [reply src/components/InviteUser.tsx]: [fix](https://github.com/PropelAuth/react/pull/11/commits#:~:text=added%20error%20message%20to%20invites)

File: src/components/Login.tsx (hunk 8 above)
Comment (joshdschneider) [reply src/components/Login.tsx]: [fix](https://github.com/PropelAuth/react/pull/11/commits#:~:text=join%20appearance%20types)

File: src/components/Login.tsx [hunk 22]
```diff
@@ -0,0 +1,217 @@
...
+    onRedirectToSignup,
+    onRedirectToForgotPassword,
+    presetEmail,
+    appearance,
+    confirmEmailAppearance,
+    verifyAppearance,
+    userMetadataAppearance,
+    createOrgAppearance,
+}: LoginProps) => {
+    const { config } = useConfig()
```
[the remark this answers is not in the document]
Comment (joshdschneider) [reply src/components/Login.tsx:67]: [fix](https://github.com/PropelAuth/react/pull/11/commits#:~:text=handle%20config%20loading%20state)

File: src/components/Login.tsx [hunk 23]
```diff
@@ -0,0 +1,217 @@
...
+    const [step, setStep] = useState<LoginStateEnum>(LoginStateEnum.LoginRequired)
```
[the remark this answers is not in the document]
Comment (joshdschneider) [reply src/components/Login.tsx]: [fix](https://github.com/PropelAuth/react/pull/11/commits#:~:text=added%20custom%20hook%20to%20manage%20login%20state)

File: src/components/Signup.tsx (hunk 5 above)
Comment (joshdschneider) [reply src/components/Signup.tsx]: [fix](https://github.com/PropelAuth/react/pull/11/commits#:~:text=697a7e1-,fix%20type%20import,-joshdschneider%20committed%20now)

File: src/components/ManageOrg.tsx [hunk 24]
```diff
@@ -0,0 +1,859 @@
...
+    options?: {
+        rowsPerPage?: number
+        editUserButtonContent?: ReactNode
+        filterButtonContent?: ReactNode
+        inviteUserButtonContent?: ReactNode
+        pageBackButtonContent?: ReactNode
+        pageNextButtonContent?: ReactNode
+        editUserModalHeaderContent?: ReactNode
+        changeRoleLabel?: ReactNode
+        saveRoleButtonContent?: ReactNode
```
[the remark this answers is not in the document]
Comment (joshdschneider) [reply src/components/ManageOrg.tsx]: [fix](https://github.com/PropelAuth/react/pull/11/commits#:~:text=better%20naming%20convention%20for%20appearance)

File: src/components/ManageOrg.tsx [hunk 25]
```diff
@@ -0,0 +1,859 @@
...
+    const [users, setUsers] = useState<User[]>([])
+    const [invitations, setInvitations] = useState<Invitation[]>([])
+    const [inviteePossibleRoles, setInviteePossibleRoles] = useState<string[]>([])
+    const [roles, setRoles] = useState<string[]>([])
+
+    useEffect(() => {
+        async function init() {
+            const response = await orgApi.selectedOrgStatus({ id: orgId })
+            if (response.ok) {
+                setUsers(response.body.users)
```
[the remark this answers is not in the document]
Comment (joshdschneider) [reply src/components/ManageOrg.tsx]: [fix](https://github.com/PropelAuth/react/pull/11/commits#:~:text=support%20promise%20cancelling)

File: src/components/ManageOrg.tsx [hunk 26]
```diff
@@ -0,0 +1,859 @@
...
+                setInvitations(response.body.pendingInvites)
+                setRoles(config?.roles || [])
+                setInviteePossibleRoles(response.body.inviteePossibleRoles)
+            }
```
[the remark this answers is not in the document]
Comment (joshdschneider) [reply src/components/ManageOrg.tsx:169]: [fix](https://github.com/PropelAuth/react/pull/11/commits#:~:text=useSelectedOrg%20error%20handling)

File: src/components/ManageOrg.tsx [hunk 27]
```diff
@@ -0,0 +1,859 @@
+import React, { ChangeEvent, Dispatch, FormEvent, ReactNode, SetStateAction, useEffect, useState } from "react"
```
[the remark this answers is not in the document]
Comment (joshdschneider) [reply src/components/ManageOrg.tsx]: yup, agreed - [fix](https://github.com/PropelAuth/react/pull/11/commits#:~:text=broke%20up%20manage%20org%20into%20multiple%20files)

File: src/components/SignInOptions.tsx (hunk 4 above)
Comment (joshdschneider) [reply src/components/SignInOptions.tsx:20]: [fix](https://github.com/PropelAuth/react/pull/11/commits#:~:text=handle%20loading%20%26%20error%20state%20for%20mfa)

File: src/components/Mfa.tsx (hunk 3 above)
Comment (joshdschneider) [reply src/components/Mfa.tsx]: [fix](https://github.com/PropelAuth/react/pull/11/commits#:~:text=handle%20loading%20%26%20error%20state%20for%20mfa)

**Pull request closed without merging**